## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [22]:
import pandas as pd

df_cds = pd.read_csv('Resources/client_dataset.csv')

df_cds.head(5)

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [23]:
# View the column names in the data

df_cds.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [24]:
# Use the describe function to gather some basic statistics

df_cds.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [25]:
# Use this space to do any additional research
# and familiarize yourself with the data.

df_cds.info()
#datatypes of each column 
df_cds.dtypes


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

first           object
last            object
job             object
phone           object
email           object
client_id        int64
order_id         int64
order_date      object
order_week       int64
order_year       int64
item_id         object
category        object
subcategory     object
unit_price     float64
unit_cost      float64
unit_weight    float64
qty              int64
line_number      int64
dtype: object

In [26]:
#find if nulls exist in the dataset 
df_cds.isnull().sum()

first          0
last           0
job            0
phone          0
email          0
client_id      0
order_id       0
order_date     0
order_week     0
order_year     0
item_id        0
category       0
subcategory    0
unit_price     0
unit_cost      0
unit_weight    0
qty            0
line_number    0
dtype: int64

In [43]:

# Used the describe function to gather some basic statistics on string columns 

print('Job Column - ') 
print(df_cds['job'].describe().to_dict().items())

print("category Column - ")
print(df_cds['category'].describe().to_dict().items())

print("subcategory Column - ")

print(df_cds['subcategory'].describe().to_dict().items())

print('Job Column unique- ') 
print(df_cds['job'].unique())

print("category Column unique- ")
print(df_cds['category'].unique())

print("subcategory Column unique- ")

print(df_cds['subcategory'].unique())



Job Column - 
dict_items([('count', 54639), ('unique', 498), ('top', 'Osteopath'), ('freq', 631)])
category Column - 
dict_items([('count', 54639), ('unique', 5), ('top', 'consumables'), ('freq', 23538)])
subcategory Column - 
dict_items([('count', 54639), ('unique', 18), ('top', 'bathroom supplies'), ('freq', 6424)])
Job Column unique- 
['Immunologist' 'Music therapist' 'Accounting technician' 'Chiropractor'
 'Production designer, theatre/television/film' 'Barista'
 'Theme park manager' 'Chiropodist' 'Secretary, company'
 'Engineer, broadcasting (operations)' 'Surveyor, land/geomatics'
 'Lobbyist' 'Psychologist, clinical' 'Research scientist (maths)'
 'Government social research officer' 'Colour technologist'
 'Production assistant, television' 'Lecturer, further education'
 'Commercial art gallery manager' 'Architectural technologist'
 'Engineer, manufacturing systems' 'Textile designer' 'Air cabin crew'
 'Medical illustrator' 'Technical brewer' 'Radiographer, therapeutic'
 'Race rel

In [28]:
# What three item categories had the most entries?
df_cds['category'].value_counts().head(3)


category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [29]:
# For the category with the most entries,
# which subcategory had the most entries?

# df_cds['category'].value_counts().idxmax() -> gives us the name of the cagegory with max count 
# df_cds[df_cds['category'] == df_cds['category'].value_counts().idxmax() ]['subcategory'] filters all rows in dataframe 
# WHERE category = max category 
#['subcategory'].value_counts() get the number of items in each group 
 

df_cds[df_cds['category'] == df_cds['category'].value_counts().idxmax() ]['subcategory'].value_counts().head(1)


subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [30]:
# Which five clients had the most entries in the data?

df_cds['client_id'].value_counts().head(5)


client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [44]:
# Store the client ids of those top 5 clients in a list.

top5bycnt=df_cds['client_id'].value_counts().head(5).index.to_list()
top5bycnt

[33615, 66037, 46820, 38378, 24741]

In [32]:
# How many total units (the qty column) did the
# client with the most entries order order?

df_cds[df_cds['client_id']==df_cds['client_id'].value_counts().idxmax()].sum()['qty']


64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [33]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty

df_cds['line_subtotal']=df_cds['unit_price']*df_cds['qty']
df_cds[['unit_price','qty','line_subtotal']].head(5)

unit_price  qty  line_subtotal
0     1096.80  105      115164.00
1       24.95   21         523.95
2       13.52   39         527.28
3       36.42   29        1056.18
4      195.10   20        3902.00

In [34]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.

# created a computed column total_weight 
df_cds['total_weight']=df_cds['unit_weight'] * df_cds['qty'] 

df_cds['shipping_price'] = df_cds.apply(lambda rw: rw['total_weight'] * 7 if rw['total_weight'] > 50 else rw['total_weight'] * 10,axis=1)

df_cds[['unit_price','unit_weight','qty','total_weight','shipping_price']].head(5)

unit_price  unit_weight  qty  total_weight  shipping_price
0     1096.80         7.50  105        787.50         5512.50
1       24.95         1.49   21         31.29          312.90
2       13.52         1.68   39         65.52          458.64
3       36.42         1.23   29         35.67          356.70
4      195.10        46.43   20        928.60         6500.20

In [35]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%

df_cds['line_price']=round((df_cds['line_subtotal'] + df_cds['shipping_price'])*(1.0925),2)

df_cds[['line_subtotal','shipping_price','line_price']].head(5)
# ** other columns exist but are not displayed 

line_subtotal  shipping_price  line_price
0      115164.00         5512.50   131839.08
1         523.95          312.90      914.26
2         527.28          458.64     1077.12
3        1056.18          356.70     1543.57
4        3902.00         6500.20    11364.40

In [46]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).

df_cds['line_cost']=round((df_cds['unit_cost']*df_cds['qty']) + df_cds['shipping_price'],2)

df_cds[['shipping_price','line_price','unit_cost','qty','line_cost']].head(5)
# ** other columns exist but are not displayed 

shipping_price  line_price  unit_cost  qty  line_cost
0         5512.50   131839.08     762.71  105   85597.05
1          312.90      914.26      15.09   21     629.79
2          458.64     1077.12       7.86   39     765.18
3          356.70     1543.57      24.85   29    1077.35
4         6500.20    11364.40     108.17   20    8663.60

In [37]:
# Create a column for the profit of
# each line using line cost and line price

df_cds['line_profit']=df_cds['line_price']-df_cds['line_cost']

df_cds[['client_id','qty','line_price','line_cost','line_profit']].head(5)



client_id  qty  line_price  line_cost  line_profit
0      58515  105   131839.08   85597.05     46242.03
1      37609   21      914.26     629.79       284.47
2      57113   39     1077.12     765.18       311.94
3      46554   29     1543.57    1077.35       466.22
4      92089   20    11364.40    8663.60      2700.80

## Part 3: Confirm your work

You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [38]:
# Check your work using the totals above

# using dataframe.query
print(f"Order ID 2742071 had a total price of ${df_cds.query('order_id==2742071')['line_price'].sum():,.2f}")

# using dataframe.loc
print(f"Order ID 2173913 had a total price of ${df_cds.loc[df_cds['order_id'] == 2173913]['line_price'].sum():,.2f}")

# using dataframe indexing 
print(f"Order ID 2173913 had a total price of ${df_cds[df_cds['order_id']==6128929]['line_price'].sum():,.2f}")


Order ID 2742071 had a total price of $152,811.89
Order ID 2173913 had a total price of $162,388.71
Order ID 2173913 had a total price of $923,441.25


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [49]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.

df_cds[df_cds['client_id'].isin(top5bycnt)].groupby(df_cds['client_id'])['line_price'].sum().sort_values().map('${:,.2f}'.format)

client_id
33615     $8,377,308.52
46820     $9,743,794.36
66037    $10,259,514.79
38378    $12,906,550.87
24741    $82,268,892.02
Name: line_price, dtype: object

In [70]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.

df_summary = df_cds[df_cds['client_id'].isin(top5bycnt)].groupby('client_id').agg({
    'qty': 'sum',
    'shipping_price': 'sum',
    'line_price': 'sum',
    'line_cost': 'sum',
    'line_profit' :'sum'
}).sort_values('line_profit',ascending=False)

df_summary


qty  shipping_price   line_price    line_cost  line_profit
client_id                                                               
24741      239862      5126448.37  82268892.02  45688899.71  36579992.31
38378       73667      3429455.40  12906550.87   9634720.98   3271829.89
66037       43018      1395151.85  10259514.79   7004482.98   3255031.81
46820       75768      1601448.84   9743794.36   7007191.64   2736602.72
33615       64313      1828984.89   8377308.52   6175313.91   2201994.61

In [75]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.
df_summary_Formatted=df_summary.apply(lambda col: col.apply(lambda cel: '${:,.2f}M'.format(cel / 10**6) if col.name !='qty' else cel ))
df_summary_Formatted.columns=['Units','Shipping','Total Revenue','Total Cost','Total Profit']
df_summary_Formatted
#'Client ID','Units','Shipping','Total Revenue','Total Cost','Total Profit'


Units Shipping Total Revenue Total Cost Total Profit
client_id                                                       
24741      239862   $5.13M       $82.27M    $45.69M      $36.58M
38378       73667   $3.43M       $12.91M     $9.63M       $3.27M
66037       43018   $1.40M       $10.26M     $7.00M       $3.26M
46820       75768   $1.60M        $9.74M     $7.01M       $2.74M
33615       64313   $1.83M        $8.38M     $6.18M       $2.20M

In [80]:
# Sort the updated data by 'Total Profit' form highest to lowest

# the data set is already sorted 
# but if it was not and  we wanted to sort again 
# we would have toconvert total profit which is a string column to a digit. 
df_summary_Formatted['ntp']= pd.to_numeric(df_summary_Formatted['Total Profit'].str.replace('$','').str.replace('M',''))

df_summary_Formatted.sort_values('ntp',ascending=False).filter(['Client ID','Units','Shipping','Total Revenue','Total Cost','Total Profit'])



Units Shipping Total Revenue Total Cost Total Profit
client_id                                                       
24741      239862   $5.13M       $82.27M    $45.69M      $36.58M
38378       73667   $3.43M       $12.91M     $9.63M       $3.27M
66037       43018   $1.40M       $10.26M     $7.00M       $3.26M
46820       75768   $1.60M        $9.74M     $7.01M       $2.74M
33615       64313   $1.83M        $8.38M     $6.18M       $2.20M